lib/widgets/login/browser_login.dart

```diff
@@ -0,0 +1,181 @@
+        otpEncryptedApiKey = otpEncryptedApiKeyStr;
+        realm = Uri.parse(realmStr);
+        userId = int.parse(userIdStr);
+        email = emailStr;
+      } else {
+        throw 'Got invalid query params from browser redirect url';
+      }
+    } catch (e, st) {
+      // TODO: Log error to Sentry
+      debugLog('$e\n$st');
+      return;
+    }
+
+    // Get the previously temporarily stored otp & serverSettings.
+    final GetServerSettingsResult serverSettings;
+    final String apiKey;
+    try {
+      final otp = _tempAuthOtp[realm];
+      _tempAuthOtp.clear();
+      final settings = _tempServerSettings[realm];
+      _tempServerSettings.clear();
+      if (otp == null) {
+        throw 'Failed to find the previously generated mobile_auth_otp';
+      }
+      if (settings == null) {
+        // TODO: Maybe try refetching instead of error-ing out.
+        throw 'Failed to find the previously stored serverSettings';
+      }
+
+      // Decode the otp XOR-ed api key
+      apiKey = _decodeApiKey(otp, otpEncryptedApiKey);
+      serverSettings = settings;
+    } catch (e, st) {
+      // TODO: Log error to Sentry
+      debugLog('$e\n$st');
+      return;
+    }
+
+    // TODO(#108): give feedback to user on SQL exception, like dupe realm+user
+    final accountId = await globalStore.insertAccount(AccountsCompanion.insert(
+      realmUrl: serverSettings.realmUri,
+      email: email,
+      apiKey: apiKey,
+      userId: userId,
+      zulipFeatureLevel: serverSettings.zulipFeatureLevel,
+      zulipVersion: serverSettings.zulipVersion,
+      zulipMergeBase: Value(serverSettings.zulipMergeBase),
+    ));
+
+    if (!context.mounted) {
+      return;
+    }
+    navigatorKey.currentState?.pushAndRemoveUntil(
+      HomePage.buildRoute(accountId: accountId),
+      (route) => (route is! LoginSequenceRoute),
+    );
+  }
+}
+
+/// Generates a `mobile_flow_otp` to be used by the server for
+/// mobile login flow, server XOR's the api key with the otp hex
+/// and returns the resulting value. So, the same otp that was passed
+/// to the server can be used again to decode the actual api key.
+String _generateMobileFlowOtp() {
+  final rand = Random.secure();
+  return hex.encode(rand.nextBytes(32));
+}
+
+String _decodeApiKey(String otp, String otpEncryptedApiKey) {
+  final otpHex = hex.decode(otp);
+  final otpEncryptedApiKeyHex = hex.decode(otpEncryptedApiKey);
+  return String.fromCharCodes(otpHex ^ otpEncryptedApiKeyHex);
+}
+
+// TODO: Remove this when upstream issue is fixed
+//  https://github.com/dart-lang/sdk/issues/53339
+extension _RandomNextBytes on Random {
+  static const int _pow2_32 = 0x100000000;
+  Uint8List nextBytes(int length) {
+    if ((length % 4) != 0) {
+      throw ArgumentError('\'length\' must be a multiple of 4');
+    }
+    final result = Uint32List(length);
+    for (int i = 0; i < length; i++) {
+      result[i] = nextInt(_pow2_32);
+    }
+    return result.buffer.asUint8List(0, length);
```

Comment:
This gets 4x as much data from the RNG as it uses, right?

I think the simpler version suggested on that upstream thread is fine, just doing `nextInt(256)`. (Well, the version there says `nextInt(255)`, but I believe that's an off-by-one bug.) It's possible that's a bit slower than a more complex version along these lines could be, but this isn't a high-performance context.